### PR TITLE
http3: Rate limit the sendmsg failed error

### PR DESCRIPTION
**Pull request by deweerdt**
In practice, these errors tend to happen all at the same time

### CODE REVIEW

**Comment by deweerdt** [comment]
Thanks for your feedback, @kazuho . What do you think of ca8a390?

File: lib/http3/common.c [hunk 1]
```diff
@@ -57,6 +57,51 @@ struct st_h2o_http3_ingress_unistream_t {
 
 const ptls_iovec_t h2o_http3_alpn[1] = {{(void *)H2O_STRLIT("h3-25")}};
 
+static struct {
+    pthread_mutex_t lock;
+    time_t last_reported;
+    size_t failures;
+    size_t successes;
```
Comment (kazuho) [inline lib/http3/common.c]: I think that the type of counters should be `uint64_t`. These counts are unrelated to the amount of memory that can be allocated, and needs to be greater than 32-bit unsigned int even in 32-bit environments.

File: lib/http3/common.c [hunk 2]
```diff
@@ -57,6 +57,51 @@ struct st_h2o_http3_ingress_unistream_t {
...
+
+static void track_sendmsg_flush(h2o_loop_t *loop)
+{
+    uint64_t now = h2o_now(loop);
+    if (track_sendmsg.last_reported + 60000 < now) {
+        /* this is racy, and might under report successes when there is a
+           failure, but allows to avoid taking the lock in the happy path */
+        if (track_sendmsg.failures == 0) {
+            track_sendmsg.last_reported = now;
+            track_sendmsg.successes = 0;
```
Comment (kazuho) [inline lib/http3/common.c]: I think it might be a good idea to define `successes` and `failures` as counters from the start of the process, rather than counters from the previous log emission. That avoids the problem of resetting the values.

When it becomes time to log, we can simply copy the values, compare them with the previous copy, and log the difference.

File: lib/http3/common.c [hunk 3]
```diff
@@ -134,8 +179,11 @@ int h2o_http3_send_datagram(h2o_http3_ctx_t *ctx, quicly_datagram_t *p)
 
         /* Temporary failure to send a packet is not a permanent error fo the connection. (TODO do we want do something more
          * specific?) */
-        perror("sendmsg failed");
+        track_sendmsg_failure();
+    } else {
+        track_sendmsg_success();
     }
+    track_sendmsg_flush(ctx->loop);
```
Comment (kazuho) [inline lib/http3/common.c]: I wonder if we need to log anything when there is no loss. To paraphrase, it might make sense to convert `track_sendmsg_flush` into a subroutine of `track_sendmsg_failure`.

The other benefit of merging `track_sendmsg_flush` into `track_sendmsg_failure` is that we can get rid of `last_errno`. `errno` would always be available when we log the message.

Comment (deweerdt) [reply lib/http3/common.c]: If we do so, we might wait for a long time for messages to appear, would you be ok with that?

Comment (kazuho) [reply lib/http3/common.c]: How about emitting a message when both of the following conditions are met:
* a new failure event occurs
* at least 60 seconds have elapsed since the last emission of the message

That would guarantee that we'd see the message at least once a minute, when loss is happening. I think that would be fine?

**Comment by kazuho** [comment]
Thank you for the changes. I've filed https://github.com/deweerdt/h2o/pull/3, please let me know what you think.

**Comment by deweerdt** [comment]
That looks good, thank you! I'm going to give it a test today

**Comment by deweerdt** [comment]
Tested and merged, thank you @kazuho . Here's a sample log:
```
Feb 13 20:46:52 localhost h2o[192687]: sendmsg failed 73 times, succeeded: 115878 times, over the last minute: Resource temporarily unavailable                               
```

**Comment by kazuho** [comment]
Good to know that the changes look okay and works.
